Java-Interfaces.md

```diff
@@ -0,0 +1,219 @@
+# Java Interfaces
+
+Interface in Java is a bit like the Class, but with a significant difference : an `interface` can _only_ have method signatures and fields. That means, an Interface cannot contain the implementation of any method, just its signature, i.e. the name, parameters and exceptions of the method. For example :
+
+```java
+public interface Vehicle {
+	public string licensePlate = "";
+	public void start();
+	public void stop();
+}
+```
+
+The interface above contains one field and two methods. Alone, it is not of much use, but they are usually used along with Classes. How? Simple, you have to make sure some class `implements` it.
+
+```java
+public class Car implements Vehicle {
+	public void start() {
+		System.out.println("starting engine...");
+	}
+	public void stop() {
+		System.out.println("stopping engine...");
+	}
+}
+```
+
+Now, there is a **ground rule** : The Class must implement **all** of the methods in the Interface. The methods must have _the exact same_ signature (name + parameters + exceptions) as described in the interface. The class _does not_ need to declare the fields though, only the methods.
+
+# 
+
+## Instances of an Interface
+
+Once you create a Java Class which `implements` any Interface, the object instance can be referenced as an instance of the Interface. Similar concept as of Inheritance instantiation.
+
+```java
+// following our previous example
+
+Vehicle tesla = new Car();
+
+tesla.start(); // starting engine ...
+```
+
+But, you **cannot** create an instance of an Interface itself. You must create instance of some class implementing an Interface to reference it. Think of interfaces as a blank contract form, or a template.
+
+What can you do with this feature? Polymorphism! You can use only interfaces to refer to object instances!
+
+```java
+class Truck implements Vehicle {
+	public void start() {
+		System.out.println("starting truck engine...");
+	}
+	public void stop() {
+		System.out.println("stopping truck engine...");
+	}
+}
+
+class Starter {
+	// static method, can be called without instantiating the class
+	public static void startEngine(Vehicle vehicle) {
+		vehicle.start();
+	}
+}
+
+Vehicle tesla = new Car();
+Vehicle tata = new Truck();
+
+Starter.startEngine(tesla); // starting engine ...
+Starter.startEngine(tata); // starting truck engine ...
+
+```
+
+## But how about multiple interfaces?
+
+Yes, you can implement multiple Interfaces in a single class. But do not forget to implement _all_ of the methods of all the Interfaces, otherwise compilation will fail!
+
+```java
+public interface GPS {
+	public LatLng getCoordinates();
+}
+
+public interface Radio {
+	public void startRadio();
+	public void stopRadio();
+}
+
+public class Smartphone implements GPS,Radio {
+	public LatLng getCoordinates() {
+		// return some coordinates
+	}
+	public void startRadio() {
+	  // start Radio
+	}
+	public void stopRadio() {
+		// stop Radio
+	}
+}
+```
+
```

Comment:
I was thinking maybe you can create some REPL forks with these code snippets, so that a reader can easily run and see these in action.

I am not sure you should do it for this very PR though...maybe in near future.